File: lib/task_index_manager.py
"""
Task Index Manager - V3.1.0
Auto-numbering and task indexing system
Ported from v2.1.0 Node.js to Python
"""

import json
from dataclasses import dataclass, field, asdict
from pathlib import Path
from typing import List, Dict, Optional, Any
from datetime import datetime


@dataclass
class TaskIndex:
    """Task index data structure"""
    version: str = "1.0.0"
    last_task_id: int = 0
    active_tasks: List[Dict[str, Any]] = field(default_factory=list)
    archived_tasks: List[Dict[str, Any]] = field(default_factory=list)
    updated_at: str = field(default_factory=lambda: datetime.now().isoformat())
    
    def to_dict(self) -> Dict[str, Any]:
        return asdict(self)
    
    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> 'TaskIndex':
        return cls(**data)


class TaskIndexManager:
    """
    Manages task indexing with auto-numbering.
    
    Features:
    - Auto-generates sequential task IDs (001, 002...)
    - Maintains active and archived task lists
    - Persists to JSON file
    """
    
    def __init__(self, index_file: Optional[str] = None):
        self.index_file = Path(index_file) if index_file else Path("task_backlog/index.json")
        self._index: Optional[TaskIndex] = None
    
    def init(self) -> None:
        """Initialize or load existing index"""
        if self.index_file.exists():
            self._load_index()
        else:
            self._create_new_index()
    
    def _create_new_index(self) -> None:
        """Create new empty index"""
        self._index = TaskIndex()
        self._save_index()
    
    def _load_index(self) -> None:
        """Load index from file"""
        try:
            with open(self.index_file, 'r', encoding='utf-8') as f:
                data = json.load(f)
            self._index = TaskIndex.from_dict(data)
        except (json.JSONDecodeError, KeyError):
            # Corrupted index, create new
            self._create_new_index()
    
    def _save_index(self) -> None:
        """Save index to file"""
        self.index_file.parent.mkdir(parents=True, exist_ok=True)
        self._index.updated_at = datetime.now().isoformat()
        with open(self.index_file, 'w', encoding='utf-8') as f:
            json.dump(self._index.to_dict(), f, indent=2)
    
    def get_next_id(self) -> int:
        """Get next sequential task ID"""
        if self._index is None:
            raise RuntimeError("Index not initialized. Call init() first.")
        
        self._index.last_task_id += 1
        self._save_index()
        return self._index.last_task_id
    
    def register_task(self, task: Dict[str, Any]) -> None:
        """Register a task in the index"""
        if self._index is None:
            raise RuntimeError("Index not initialized. Call init() first.")
        
        # Check if task already exists
        existing = self.get_task(task.get("id"))
        if existing:
            # Update existing
            for i, t in enumerate(self._index.active_tasks):
                if t.get("id") == task["id"]:
                    self._index.active_tasks[i] = task
                    break
        else:
            # Add new
            self._index.active_tasks.append(task)
        
        self._save_index()
    
    def get_task(self, task_id: str) -> Optional[Dict[str, Any]]:
        """Get task by ID"""
        if self._index is None:
            return None
        
        for task in self._index.active_tasks:
            if task.get("id") == task_id:
                return task
        
        # Also check archived
        for task in self._index.archived_tasks:
            if task.get("id") == task_id:
                return task
        
        return None
    
    def list_active(self) -> List[Dict[str, Any]]:
        """List all non-completed active tasks"""
        if self._index is None:
            return []
        
        return [
            task for task in self._index.active_tasks
            if task.get("status") != "completed"
        ]
    
    def archive_task(self, task_id: str) -> bool:
        """Move task to archive"""
        if self._index is None:
            return False
        
        for i, task in enumerate(self._index.active_tasks):
            if task.get("id") == task_id:
                task["archived_at"] = datetime.now().isoformat()
                self._index.archived_tasks.append(task)
                self._index.active_tasks.pop(i)
                self._save_index()
                return True
        
        return False
    
    def get_stats(self) -> Dict[str, int]:
        """Get index statistics"""
        if self._index is None:
            return {"total": 0, "active": 0, "archived": 0}
        
        return {
            "total": self._index.last_task_id,
            "active": len(self._index.active_tasks),
            "archived": len(self._index.archived_tasks)
        }


# Convenience function for V3 integration
def format_task_id(number: int) -> str:
    """Format number as task ID (001, 002...)"""
    return f"task_{number:03d}"

File: lib/task-index-manager.js
/**
 * Task Index Manager - Manages active task list and ID assignment
 * Part of task-workflow archive system
 */

const fs = require('fs').promises;
const path = require('path');

class TaskIndexManager {
  constructor(options = {}) {
    this.config = {
      activeDir: options.activeDir || './01_active/tasks',
      archiveDir: options.archiveDir || './02_archive',
      indexFile: options.indexFile || './01_active/tasks/index.json',
      ...options
    };
    this.index = null;
  }

  /**
   * Initialize the index manager
   */
  async init() {
    await this.ensureDirectories();
    await this.loadIndex();
  }

  /**
   * Ensure directory structure exists
   */
  async ensureDirectories() {
    const dirs = [
      this.config.activeDir,
      path.dirname(this.config.indexFile)
    ];

    for (const dir of dirs) {
      await fs.mkdir(dir, { recursive: true });
    }
  }

  /**
   * Load or create index
   */
  async loadIndex() {
    try {
      const content = await fs.readFile(this.config.indexFile, 'utf-8');
      this.index = JSON.parse(content);
    } catch {
      // Create new index
      this.index = {
        version: '1.0.0',
        lastTaskId: 0,
        activeTasks: [],
        archivedTasks: [],
        updatedAt: new Date().toISOString()
      };
      await this.saveIndex();
    }
  }

  /**
   * Save index to file
   */
  async saveIndex() {
    this.index.updatedAt = new Date().toISOString();
    await fs.writeFile(
      this.config.indexFile,
      JSON.stringify(this.index, null, 2)
    );
  }

  /**
   * Create a new task
   */
  async createTask(taskData) {
    const taskId = this.index.lastTaskId + 1;
    this.index.lastTaskId = taskId;

    const taskNumber = String(taskId).padStart(3, '0');
    const taskName = taskData.name.toLowerCase().replace(/\s+/g, '_');
    const filename = `task_${taskNumber}_${taskName}.md`;

    const task = {
      id: taskId,
      number: taskNumber,
      name: taskData.name,
      filename,
      description: taskData.description || '',
      status: 'pending',
      priority: taskData.priority || 'medium',
      createdAt: new Date().toISOString(),
      updatedAt: new Date().toISOString(),
      filepath: path.join(this.config.activeDir, filename)
    };

    // Add to active tasks
    this.index.activeTasks.push(task);
    await this.saveIndex();

    // Create task file from template
    await this.createTaskFile(task, taskData);

    return task;
  }

  /**
   * Create task file from template
   */
  async createTaskFile(task, taskData) {
    const template = `# Task ${task.number}: ${task.name}

**Created:** ${new Date().toISOString().split('T')[0]}  
**Status:** ${task.status}  
**Priority:** ${task.priority}  
**Task ID:** ${task.number}

---

## Description

${taskData.description || 'No description provided.'}

## Objectives

${taskData.objectives ? taskData.objectives.map(o => `- [ ] ${o}`).join('\n') : '- [ ] Define objectives'}

## Progress

| Phase | Status | Started | Completed |
|-------|--------|---------|-----------|
| 1 - Planning | ⏳ | - | - |
| 2 - Execution | ⏳ | - | - |
| 3 - Validation | ⏳ | - | - |

## Notes

${taskData.notes || ''}

---

*Created by task-workflow 🜁*
`;

    await fs.writeFile(task.filepath, template);
  }

  /**
   * Get all active tasks
   */
  getActiveTasks() {
    return this.index.activeTasks.filter(t => t.status !== 'completed');
  }

  /**
   * Get a specific task by ID
   */
  getTask(taskId) {
    return this.index.activeTasks.find(t => t.id === taskId);
  }

  /**
   * Update task status
   */
  async updateTaskStatus(taskId, status) {
    const task = this.getTask(taskId);
    if (!task) {
      throw new Error(`Task ${taskId} not found`);
    }

    task.status = status;
    task.updatedAt = new Date().toISOString();

    await this.saveIndex();
    return task;
  }

  /**
   * Archive a completed task
   */
  async archiveTask(taskId) {
    const taskIndex = this.index.activeTasks.findIndex(t => t.id === taskId);
    if (taskIndex === -1) {
      throw new Error(`Task ${taskId} not found`);
    }

    const task = this.index.activeTasks[taskIndex];

    if (task.status !== 'completed') {
      throw new Error(`Task ${taskId} is not completed (status: ${task.status})`);
    }

    // Determine archive location
    const archiveDate = new Date();
    const yearMonth = `${archiveDate.getFullYear()}-${String(archiveDate.getMonth() + 1).padStart(2, '0')}`;
    const archiveDir = path.join(this.config.archiveDir, yearMonth, 'tasks');
    await fs.mkdir(archiveDir, { recursive: true });

    // Update filename for archive
    const archivedFilename = task.filename.replace('.md', '_completed.md');
    const archivePath = path.join(archiveDir, archivedFilename);

    // Move file
    await fs.rename(task.filepath, archivePath);

    // Update task record
    task.archivedAt = new Date().toISOString();
    task.originalPath = task.filepath;
    task.archivePath = archivePath;

    // Move from active to archived
    this.index.activeTasks.splice(taskIndex, 1);
    this.index.archivedTasks.push(task);

    await this.saveIndex();

    return {
      ...task,
      archiveLocation: archivePath
    };
  }

  /**
   * Generate task status report
   */
  async generateReport() {
    const active = this.getActiveTasks();
    const pending = active.filter(t => t.status === 'pending');
    const inProgress = active.filter(t => t.status === 'in_progress');

    return {
      summary: {
        totalActive: active.length,
        pending: pending.length,
        inProgress: inProgress.length,
        completed: this.index.archivedTasks.length
      },
      pendingTasks: pending,
      inProgressTasks: inProgress,
      recentArchived: this.index.archivedTasks.slice(-5)
    };
  }
}

module.exports = TaskIndexManager;
